javaslang/src/main/java/javaslang/collection/TreeMap.java

```diff
@@ -1225,34 +1183,201 @@ public String toString() {
+        }
+        return new TreeMap<>(tree);
+    }
+
+    @SafeVarargs
+    private static <K, V> TreeMap<K, V> createFromMapEntries(EntryComparator<K, V> entryComparator, java.util.Map.Entry<? extends K, ? extends V>... entries) {
+        Objects.requireNonNull(entries, "entries is null");
+        RedBlackTree<Tuple2<K, V>> tree = RedBlackTree.empty(entryComparator);
+        for (java.util.Map.Entry<? extends K, ? extends V> entry : entries) {
+            final K key = entry.getKey();
+            final V value = entry.getValue();
+            tree = tree.insert(Tuple.of(key, value));
+        }
+        return new TreeMap<>(tree);
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <K, V> TreeMap<K, V> createFromPairs(EntryComparator<K, V> entryComparator, Object... pairs) {
+        RedBlackTree<Tuple2<K, V>> tree = RedBlackTree.empty(entryComparator);
+        for (int i = 0; i < pairs.length; i += 2) {
+            final K key = (K) pairs[i];
+            final V value = (V) pairs[i + 1];
+            tree = tree.insert(Tuple.of(key, value));
+        }
+        return new TreeMap<>(tree);
+    }
+
     private TreeMap<K, V> createFromEntries(Iterable<Tuple2<K, V>> tuples) {
-        return createTreeMap(entries.comparator(), tuples);
+        return createTreeMap((EntryComparator<K, V>) entries.comparator(), tuples);
     }
 
     private TreeMap<K, V> emptyInstance() {
         return isEmpty() ? this : new TreeMap<>(entries.emptyInstance());
     }
 
-    /**
-     * Used to compare entries by key and store the keyComparator for later access.
-     *
-     * @param <K> key type
-     * @param <V> value type, needed at compile time for the Comparator interface
-     */
-    static class EntryComparator<K, V> implements Comparator<Tuple2<K, V>>, Serializable {
+    // -- internal types
+
+    private interface EntryComparator<K, V> extends Comparator<Tuple2<K, V>>, Serializable {
 
-        private static final long serialVersionUID = 1L;
+        long serialVersionUID = 1L;
 
-        final Comparator<K> keyComparator;
+        static <K, V> EntryComparator<K, V> of(Comparator<? super K> keyComparator) {
+            Objects.requireNonNull(keyComparator, "keyComparator is null");
+            return new Specific<>(keyComparator);
+        }
 
-        @SuppressWarnings("unchecked")
-        EntryComparator(Comparator<? super K> keyComparator) {
-            this.keyComparator = (Comparator<K>) keyComparator;
+        static <K, V> EntryComparator<K, V> natural() {
+            return Natural.instance();
         }
 
-        @Override
-        public int compare(Tuple2<K, V> e1, Tuple2<K, V> e2) {
-            return keyComparator.compare(e1._1, e2._1);
+        Comparator<K> keyComparator();
+
+        // -- internal impls
+
+        final class Specific<K, V> implements EntryComparator<K, V> {
+
+            private static final long serialVersionUID = 1L;
+
+            final Comparator<K> keyComparator;
```

Comment:
\[Checkstyle\] ERROR: Variable 'keyComparator' must be private and have accessor methods\.